### ISSUE TITLE
Clarification about the paper and Code for Uncertainty Query Selection

### ISSUE THREAD

**Issue by WassimBouatay** [lyuwenyu/RT-DETR#605]
1- The AIFI model sounds like just the normal encoder of the original DETR, but I understand the contribution of the paper is the addition CCFF. But how is the RT-DETR faster than DETR in that case ? I understand it is faster than DINO or Deformable DETR for doing a single Scale Attention, but it was also the case for the original DETR. 

2- I don't seem to find the code for the Uncertainty Query selection. What is P(X) and C(X) ? There is no explanation in the paper about them ? I saw many issues discussion the vfl loss, but it seems to me that just replace the target class from 1 to IOU value using the predicted boxes, instead of using the formula of the paper. Is that it ? 

```
target_score_o = torch.zeros_like(target_classes, dtype=src_logits.dtype)
target_score_o[idx] = ious.to(target_score_o.dtype)
target_score = target_score_o.unsqueeze(-1) * target
```

**Comment by SebastianJanampa** [comment]
> 1- The AIFI model sounds like just the normal encoder of the original DETR, but I understand the contribution of the paper is the addition CCFF. But how is the RT-DETR faster than DETR in that case ? I understand it is faster than DINO or Deformable DETR for doing a single Scale Attention, but it was also the case for the original DETR.

DETR uses a transformer encoder that has six decoder layers inside. That means you perform the attention mechanism 6 times. Additionally, DETR uses a single scale, and, for inference, DETR resizes the image such that the shortest side is equal to 800pix (it keeps the image ratio). 
On the other hand, RT-DETR uses square images (640x640 during inference). RT-DETR's proposed hybrid encoder applies the self-attention mechanism to the smallest feature map, and propagates the information to the other feature maps using convolution (This is also called cross-scale interaction). 


> 2- I don't seem to find the code for the Uncertainty Query selection. What is P(X) and C(X) ? There is no explanation in the paper about them ? I saw many issues discussion the vfl loss, but it seems to me that just replace the target class from 1 to IOU value using the predicted boxes, instead of using the formula of the paper. Is that it?
The quick answer is yes. Notice that now the classification loss will depend on the iou. Nevertheless, it is important to notice that you detach the iou parameters.
https://github.com/lyuwenyu/RT-DETR/blob/a80dee362b5444c1af3e8bfcb965a8892112bfe2/rtdetr_pytorch/src/zoo/rtdetr/rtdetr_criterion.py#L115-L118

**Comment by WassimBouatay** [comment]
Thank you for your reply @SebastianJanampa. 

I still have doubts about the computational time.

1- RT-DETR uses a single scale for the encoder, and so does DETR, so there is no gain in the encoder computational cost. 
2- RT-DETR has an additional module CCFF, altough it is not computationally expensive it should still add some overhead
3- For the decoder both RT-DETR and DETR have 6 layers. But RT-DETR uses deformable attention unlike DETR. 

So my understanding is that:
*  RT-DETR is faster than DETR in the decoder part for using the deformable attention + lower resolution 640 instead of 800. 
*  RT-DETR is faster than Deformable DETR (and DINO) thanks to the gains in encoder of only processing one scale + lower resolution 640 instead of 800. But I am still surprised to see RT-DETR being 20 times faster than DINO.

**Comment by SebastianJanampa** [comment]
> 1- RT-DETR uses a single scale for the encoder, and so does DETR, so there is no gain in the encoder computational cost. 2- RT-DETR has an additional module CCFF, altough it is not computationally expensive it should still add some overhead 3- For the decoder both RT-DETR and DETR have 6 layers. But RT-DETR uses deformable attention unlike DETR.

RT-DETR uses a one-layer decoder.  

> 2- RT-DETR has an additional module CCFF, altough it is not computationally expensive it should still add some overhead

Yes, it does. If you're interested in how to speed-up the encoder, you can also read [LITE-DETR](https://arxiv.org/pdf/2303.07335)
which uses the transformer-encoder from DINO, but it reduces DINO latency by over 80%, and the AP drops only 2%

> 3- For the decoder both RT-DETR and DETR have 6 layers. But RT-DETR uses deformable attention, unlike DETR.
It's not only that. DETR uses 100 queries while RT-DETR uses 300 queries. Yes, in this case, using the deformable attention significantly reduces the time complexity of RT-DETR. However, there are other factors, such as the evaluation size. RT-DETR resizes images to `640x640`, while DETR resizes them such that their shortest side is equal to 800 and the longest is not greater than 1333 (DETR tries to keep the image aspect ratio)


> So my understanding is that:
> 
> * RT-DETR is faster than DETR in the decoder part for using the deformable attention + lower resolution 640 instead of 800.

Correct.

> * RT-DETR is faster than Deformable DETR (and DINO) thanks to the gains in encoder of only processing one scale + lower resolution 640 instead of 800. But I am still surprised to see RT-DETR being 20 times faster than DINO.
The reported FPS uses tensor and half-precision. Tensorrt is a library that optimizes your model to speed up inference up to 5x faster (if you use half precision).

**Comment by talcs** [comment]
So in `vfl_loss` the ious are multiplied by the classification scores? But what is done at inference, as you have no GT and cannot compute IoUs? How are the best encoder features picked?
